File: apps/ios/WindmillKit/Tests/WindmillGymTests/UndoWindowTests.swift
import XCTest
@testable import WindmillGym
@testable import WindmillPlatform

@MainActor
final class UndoWindowTests: XCTestCase {
    private var queueURL: URL!
    private var catalogURL: URL!
    private var clockMs: Int64 = 1_000

    override func setUp() async throws {
        queueURL = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("gym-undo-\(UUID().uuidString).json")
        catalogURL = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("gym-undo-catalog-\(UUID().uuidString).json")
        clockMs = 1_000
    }

    override func tearDown() async throws {
        try? FileManager.default.removeItem(at: queueURL)
        try? FileManager.default.removeItem(at: catalogURL)
    }

    private func queueOnDisk(of seat: String? = "u1") -> SetQueue {
        let held = SetQueue(url: queueURL, deviceHolds: nil)
        held.open(under: seat)
        return held
    }

    private func liveStore(_ server: FakeTraining) async -> TrainingStore {
        server.open(Session(id: "ses_1", startedAtMs: 1_000))
        let store = TrainingStore(queue: SetQueue(url: queueURL, deviceHolds: nil),
                                  deviceCatalog: DeviceCatalog(url: catalogURL),
                                  accountCopy: AccountCopy(url: catalogURL.appendingPathExtension("account")),
                                  localLog: LocalLog(url: catalogURL.appendingPathExtension("local"), deviceHolds: nil),
                                  now: { self.clockMs },
                                  mintSession: { "ses_1" },
                                  mintSet: { "set_\(self.clockMs)" },
                                  sync: { _ in server })
        await store.connect(to: Account(
            api: WindmillApi(baseURL: URL(string: "https://windmill.works")!, credential: { nil }),
            user: User(id: "u1", email: "sam@example.com", name: "Sam")
        ))
        await store.choose("bench-press")
        return store
    }

    func testASetJustLoggedIsOnTheDeviceAndNotYetOnTheLog() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)

        XCTAssertEqual(store.sets.map(\.weightKg), [82.5])
        XCTAssertEqual(server.appended.count, 0, "nothing goes out while it can still be taken back")
        XCTAssertEqual(store.undoable?.weightKg, 82.5)
        XCTAssertEqual(queueOnDisk().pending.count, 1, "and it is on disk, not in memory")
    }

    func testASetInsideItsWindowIsNotCalledOffline() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)

        XCTAssertEqual(store.saveState, .idle)
        XCTAssertNil(store.saveState.line)
    }

    func testUndoTakesTheSetBackAndTheLogNeverHearsOfIt() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)
        XCTAssertTrue(store.undoLast())

        XCTAssertEqual(store.sets, [])
        XCTAssertEqual(server.appended.count, 0)
        XCTAssertEqual(queueOnDisk().pending.count, 0, "and the disk agrees")
        XCTAssertNil(store.undoable)
    }

    func testUndoTakesBackTheNewestSetAndLeavesTheRest() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)
        clockMs += 1_000
        await store.logSet(weightKg: 85, reps: 3)
        XCTAssertTrue(store.undoLast())

        XCTAssertEqual(store.sets.map(\.weightKg), [82.5])
    }

    func testWhenTheWindowClosesTheSetGoesOutByItself() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)
        clockMs += SetQueue.undoWindowMs + 1
        await store.flushPendingSets()

        XCTAssertEqual(server.appended.map(\.weightKg), [82.5])
        XCTAssertEqual(store.saveState, .onTheLog)
        XCTAssertNil(store.undoable, "and there is nothing left to take back")
    }

    func testUndoAfterTheSetHasLandedRefusesRatherThanPretending() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)
        clockMs += SetQueue.undoWindowMs + 1
        await store.flushPendingSets()

        XCTAssertFalse(store.undoLast())
        XCTAssertEqual(store.sets.map(\.weightKg), [82.5], "the set the log has stays on screen")
    }

    func testFinishSendsASetStillInsideItsWindowBeforeItCloses() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 100, reps: 5)
        guard case .closed = await store.finish() else {
            return XCTFail("the session closed, because nothing was left to lose")
        }

        XCTAssertEqual(server.appended.map(\.weightKg), [100])
        let order = server.calls.filter { $0 == "append" || $0 == "finish" }
        XCTAssertEqual(order, ["append", "finish"], "the set goes out before the close, never after")
    }

    // Leaving keeps the window. There is no forced flush left to end one early: what is held stays
    // held on the queue's own on-disk clock, and only that clock sends it.
    func testLeavingTheRoomKeepsTheWindowAndSendsNothingEarly() async {
        let server = FakeTraining()
        let store = await liveStore(server)

        await store.logSet(weightKg: 82.5, reps: 5)
        await store.flushPendingSets()
        XCTAssertEqual(server.appended.count, 0, "leaving the room keeps the window")
        XCTAssertEqual(store.undoable?.weightKg, 82.5, "and the way back is still open")

        clockMs += SetQueue.undoWindowMs + 1
        await store.flushPendingSets()
        XCTAssertEqual(server.appended.map(\.weightKg), [82.5])
    }

    func testAReadOfTheLogSendsWhatIsHeldFirst() async {
        let server = FakeTraining()
        let store = await liveStore(server)
        await store.logSet(weightKg: 82.5, reps: 5)

        let relaunched = TrainingStore(queue: SetQueue(url: queueURL, deviceHolds: nil),
                                       deviceCatalog: DeviceCatalog(url: catalogURL),
                                       accountCopy: AccountCopy(url: catalogURL.appendingPathExtension("account")),
                                       localLog: LocalLog(url: catalogURL.appendingPathExtension("local"), deviceHolds: nil),
                                       now: { self.clockMs },
                                       mintSession: { "ses_1" },
                                       mintSet: { "set_x" },
                                       sync: { _ in server })
        await relaunched.connect(to: Account(
            api: WindmillApi(baseURL: URL(string: "https://windmill.works")!, credential: { nil }),
            user: User(id: "u1", email: "sam@example.com", name: "Sam")
        ))

        XCTAssertEqual(server.appended.map(\.weightKg), [82.5])
        _ = store
    }
}

File: apps/android/gym/src/test/kotlin/works/windmill/gym/store/UndoWindowTests.kt
package works.windmill.gym.store

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import works.windmill.gym.domain.Session
import works.windmill.gym.domain.TrainingSet

class UndoWindowTests {
    @get:Rule
    val tmp = TemporaryFolder()

    private var clockMs = 1_000L
    private lateinit var file: File
    private lateinit var queue: SetQueue

    @Before
    fun setUp() {
        clockMs = 1_000L
        file = File(tmp.root, "gym-undo-${System.nanoTime()}.json")
        queue = SetQueue(file) { clockMs }
        queue.hold(Session(id = "ses_1", startedAtMs = 1_000))
    }

    private fun logSet(id: String, weightKg: Double = 82.5, reps: Int = 5) {
        queue.store(
            TrainingSet(id = id, exerciseId = "bench-press", weightKg = weightKg, reps = reps,
                completedAtMs = clockMs),
            sessionId = "ses_1", needsPush = true, heldUntilMs = clockMs + SetQueue.undoWindowMs)
        queue.flush()
    }

    @Test
    fun testASetJustLoggedIsOnTheDeviceAndNotYetOnTheLog() {
        logSet("set_1")

        assertEquals(listOf(82.5), queue.sets.map { it.weightKg })
        assertNull("nothing goes out while it can still be taken back",
            queue.nextOwed(skipping = emptySet(), readyAt = clockMs))
        assertEquals(82.5, queue.withdrawable()?.set?.weightKg)
        assertEquals("and it is on disk, not in memory", 1, SetQueue(file).pending.size)
    }

    @Test
    fun testASetInsideItsWindowIsNotOfferedToTheWalk() {
        logSet("set_1")

        assertNull(queue.nextOwed(skipping = emptySet(), readyAt = clockMs))
        assertEquals("held on purpose is not stranded",
            emptyList<SetQueue.Entry>(), queue.pending.filter { !it.isHeld(clockMs) })
    }

    @Test
    fun testUndoTakesTheSetBackAndTheLogNeverHearsOfIt() {
        logSet("set_1")
        assertTrue(queue.withdraw("set_1"))
        queue.flush()

        assertEquals(emptyList<TrainingSet>(), queue.sets)
        assertEquals("and the disk agrees", 0, SetQueue(file).pending.size)
        assertNull(queue.withdrawable())
    }

    @Test
    fun testUndoTakesBackTheNewestSetAndLeavesTheRest() {
        logSet("set_1")
        clockMs += 1_000
        logSet("set_2", weightKg = 85.0, reps = 3)

        val newest = queue.withdrawable()
        assertEquals(85.0, newest?.set?.weightKg)
        assertTrue(queue.withdraw(newest!!.set.id))
        assertEquals(listOf(82.5), queue.sets.map { it.weightKg })
    }

    @Test
    fun testWhenTheWindowClosesTheSetIsOfferedByItself() {
        logSet("set_1")
        clockMs += SetQueue.undoWindowMs + 1

        assertEquals("set_1", queue.nextOwed(skipping = emptySet(), readyAt = clockMs)?.set?.id)
        assertNull("and there is nothing left to take back", queue.withdrawable())
    }

    @Test
    fun testUndoAfterTheSetHasLandedRefusesRatherThanPretending() {
        logSet("set_1")
        queue.delivered(
            TrainingSet(id = "set_1", exerciseId = "bench-press", setNumber = 1, weightKg = 82.5,
                reps = 5, completedAtMs = clockMs),
            id = "set_1", sessionId = "ses_1")

        assertFalse(queue.withdraw("set_1"))
        assertEquals("the set the log has stays on screen",
            listOf(82.5), queue.sets.map { it.weightKg })
        assertNull(queue.withdrawable())
    }

    @Test
    fun testAForcedWalkOffersASetStillInsideItsWindow() {
        logSet("set_1")
        assertNull("the ordinary walk keeps the window",
            queue.nextOwed(skipping = emptySet(), readyAt = clockMs))

        val offered = queue.nextOwed(skipping = emptySet(), readyAt = null)
        assertEquals("set_1", offered?.set?.id)

        queue.delivered(offered!!.set.copy(setNumber = 1), id = "set_1", sessionId = "ses_1")
        assertEquals("nothing is left to strand against the close",
            emptyList<SetQueue.Entry>(), queue.owed("ses_1"))
        queue.close("ses_1")
        assertTrue(queue.pending.isEmpty())
    }

    @Test
    fun testARelaunchStillOwesTheSetAndAForcedWalkOffersIt() {
        logSet("set_1")

        val relaunched = SetQueue(file) { clockMs }
        assertEquals(listOf("set_1"), relaunched.pending.map { it.set.id })
        assertEquals("set_1", relaunched.nextOwed(skipping = emptySet(), readyAt = null)?.set?.id)
    }

    @Test
    fun testAQueueFileWrittenBeforeTheWindowExistedStillOpens() {
        val old = File(tmp.root, "gym-old-${System.nanoTime()}.json")
        old.writeText(
            """{"session":{"id":"ses_1","startedAt":1000},"entries":{"set_a":{"set":""" +
                """{"id":"set_a","exerciseId":"bench-press","weightKg":82.5,"reps":5,""" +
                """"completedAt":1100},"sessionId":"ses_1","needsPush":true,"remints":0}}}""")

        val opened = SetQueue(old, "u1") { clockMs }
        assertEquals("ses_1", opened.session?.id)
        assertEquals(listOf("set_a"), opened.pending.map { it.set.id })
        assertFalse("no key, no hold", opened.pending.single().isHeld(at = 0))
        assertNotNull("and the ordinary walk offers it at once",
            opened.nextOwed(skipping = emptySet(), readyAt = clockMs))
        assertEquals(emptyList<String>(), opened.order)
    }
}
